### PR TITLE
Support cancelling image copy operation

### PR DESCRIPTION
**Pull request by albertodonato**
Cancellation is not wired in the client yet, but it's possible to perform it via direct API call.

An operation is now considered cancellable when the .Cancellable(true) is called, so that the cancel channel is set.

### CODE REVIEW

File: lxd/images.go [hunk 1]
```diff
@@ -720,35 +720,25 @@ func imagesPost(d *Daemon, r *http.Request) Response {
...
 				imagePublishLock.Unlock()
 			}
-		} else {
-			/* Processing image upload */
-			info, err = getImgPostInfo(d, r, builddir, post)
-			if err != nil {
-				return err
-			}
+		}
+		if err != nil {
```
Comment (brauner) [inline lxd/images.go]: This is more of a reminder than doubt: Did you check that the scoping is correct. Go's variable scoping really really freaks me out. :)

File: lxd/operations.go [hunk 2]
```diff
@@ -153,17 +154,18 @@ func (op *operation) Cancel() (chan error, error) {
 		return nil, fmt.Errorf("Only running operations can be cancelled")
 	}
 
+	op.lock.Lock()
 	if !op.mayCancel() {
+		op.lock.Unlock()
```
Comment (brauner) [inline lxd/operations.go]: Why didn't we have the lock before and why do we need it now? (Just curious?)

File: lxd/images.go (hunk 1 above)
Comment (albertodonato) [inline lxd/images.go]: mhm, it seems it was wrong before too? err is not declared locally in the run function, but it should. I'll fix it

Comment (brauner) [inline lxd/images.go]: Thanks!

Comment (albertodonato) [inline lxd/images.go]: Done.

File: lxd/operations.go (hunk 2 above)
Comment (albertodonato) [inline lxd/operations.go]: Before, the Lock() was performed later (see the negative change), so this branch was not run with lock.
The reason I moved it up is that mayCancel() now looks at the channel, and there's a chance for a race condition if it returns true (because the channel is not nil) but before it locks another goroutine unsets it (by calling Cancellable(false)).

Comment (brauner) [inline lxd/operations.go]: Thanks!

**Comment by brauner** [comment]
lgtm, but I want @stgraber to have a look at this as well. :)

**Comment by brauner** [comment]
Well, once Jenkins is happy of course. :)

File: test/suites/image.sh [hunk 3]
```diff
@@ -48,5 +48,4 @@ test_image_list_all_aliases() {
     # both aliases are listed if the "aliases" column is included in output
     lxc image list -c L | grep -q testimage
     lxc image list -c L | grep -q zzz
-
```
Comment (stgraber) [inline test/suites/image.sh]: Can you undo this change?

Comment (albertodonato) [inline test/suites/image.sh]: (un)done
